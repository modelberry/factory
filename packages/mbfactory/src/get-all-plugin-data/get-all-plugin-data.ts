import ts from 'typescript'
import chalk from 'chalk'
import {
  parseModelberryInterface,
  ModelberryInterface,
} from '../modelberry-parser/parse-modelberry-interface'
import {
  parseModelberryVariable,
  ModelberryVariable,
} from '../modelberry-parser/parse-modelberry-variable'

export type ModelberryType = {
  interface: ModelberryInterface
  variables: ModelberryVariable[]
}

export type TypeData = {
  [typeName: string]: ModelberryType
}

export type PluginData = {
  types: TypeData
  dataVar: ModelberryVariable
}

export type AllPluginData = {
  [pluginName: string]: PluginData
}

export interface GetPluginData {
  program: ts.Program
}

export const getAllPluginData = ({ program }: GetPluginData) => {
  const allPluginData: AllPluginData = {}
  const checker = program.getTypeChecker()
  const log = console.log

  let dataVar = {}
  for (const sourceFile of program.getSourceFiles()) {
    if (sourceFile.isDeclarationFile) continue
    log(chalk.underline(sourceFile.fileName))
    const wrInterfaces: ModelberryInterface[] = []
    const modelberryVariables: ModelberryVariable[] = []

    ts.forEachChild(sourceFile, (node: ts.Node) => {
      const wrInterface = parseModelberryInterface({ node, checker })
      if (wrInterface) wrInterfaces.push(wrInterface)
      const mbVariable = parseModelberryVariable({ node, sourceFile })
      if (mbVariable.name === 'modelberryPluginData') {
        dataVar = mbVariable
      } else if (
        mbVariable.name &&
        mbVariable.isArray &&
        mbVariable.isExported
      ) {
        modelberryVariables.push(mbVariable)
      } else if (mbVariable.name && !mbVariable.isArray) {
        log(chalk.blue(`- content variable, not an array: ${mbVariable.name}`))
      } else {
        if (mbVariable.name && !mbVariable.isExported)
          log(
            chalk.blue(`- content variable, not exported: ${mbVariable.name}`)
          )
      }
    })
    // Order interfaces by interface and type
    for (const wrInterface of wrInterfaces) {
      const pluginName = (wrInterface?.interfaceTags || {})['@plugin']
      if (!pluginName) continue
      if (!wrInterface.typeName) continue
      if (!allPluginData[pluginName])
        allPluginData[pluginName] = {
          types: {},
          dataVar: {},
        }
      if (!allPluginData[pluginName].types[wrInterface.typeName]) {
        allPluginData[pluginName].types[wrInterface.typeName] = {
          interface: {},
          variables: [],
        }
      }

      allPluginData[pluginName].types[wrInterface.typeName].interface =
        wrInterface
    }
    // Merge in variables
    for (const mbVariable of modelberryVariables) {
      let lookupSuccess = false
      for (const pluginData of Object.values(allPluginData)) {
        const interfaceType = mbVariable.type || '<none>'
        if (Object.keys(pluginData.types).includes(interfaceType)) {
          pluginData.types[interfaceType].variables.push(mbVariable)
          lookupSuccess = true
        }
      }
      if (!lookupSuccess) {
        log(
          chalk.blue(
            `- content variable, no type match: ${mbVariable.type}/${mbVariable.name}`
          )
        )
      }
    }
  }
  // Add dataVar to all plugins
  Object.values(allPluginData).forEach(
    (pluginData) => (pluginData.dataVar = dataVar)
  )
  return allPluginData
}
