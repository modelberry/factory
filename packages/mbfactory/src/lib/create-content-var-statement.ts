import ts from 'typescript'

export interface CreateContentVarStatement {
  contentArray: any[]
}

export const createContentVarStatement = ({
  contentArray,
}: CreateContentVarStatement) => {
  const contentString = JSON.stringify(contentArray)
  const file = ts.createSourceFile(
    'dummy.ts',
    `export const myTopics: ContentfulTopic[] = ${contentString}`,
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS
  )
  return file
}
