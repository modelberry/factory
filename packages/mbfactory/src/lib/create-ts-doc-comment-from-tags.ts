export type InlineTags = {
  [tagName: string]: string
}

export interface CreateTsDocCommentFromTags {
  inlineTags: InlineTags
  blockTag: string
}

export const createTsDocCommentFromTags = ({
  blockTag,
  inlineTags,
}: CreateTsDocCommentFromTags) => {
  let comment = ``
  comment += `${blockTag}\n`
  for (const inlineTag in inlineTags) {
    comment += `* - {${inlineTag} ${inlineTags[inlineTag]}}\n`
  }
  return comment
}
