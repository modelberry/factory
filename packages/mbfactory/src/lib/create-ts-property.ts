import ts from 'typescript'

export type PropertryNode = {
  comment?: string
  isRequired?: boolean
  syntaxKind?: ts.SyntaxKind
}

export type PropertyTree = {
  [propertyName: string]: {
    node?: PropertryNode
    edges?: PropertyTree
  }
}

export interface CreateTsProperty {
  propertyTree: PropertyTree
}

export const createTsProperty = ({ propertyTree }: CreateTsProperty) => {
  const members = []
  for (const propertyName in propertyTree) {
    const { node, edges } = propertyTree[propertyName]
    let childNode: ts.TypeNode | undefined
    if (edges) {
      childNode = ts.factory.createTypeLiteralNode(
        createTsProperty({ propertyTree: edges })
      )
    } else if (node) {
      childNode = ts.factory.createKeywordTypeNode(
        node.syntaxKind as ts.KeywordTypeSyntaxKind
      )
    }
    if (childNode) {
      const property = ts.factory.createPropertySignature(
        undefined, // modifier
        propertyName,
        node?.isRequired
          ? undefined
          : ts.factory.createToken(ts.SyntaxKind.QuestionToken),
        childNode
      )
      if (node?.comment) {
        ts.addSyntheticLeadingComment(
          property,
          ts.SyntaxKind.MultiLineCommentTrivia,
          `${node.comment}`,
          true
        )
      }

      members.push(property)
    }
  }
  return members
}
