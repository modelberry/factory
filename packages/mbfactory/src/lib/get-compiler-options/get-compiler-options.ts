import ts from 'typescript'

export const getCompilerOptions = () => {
  const configFileName = ts.findConfigFile(
    './',
    ts.sys.fileExists,
    // TODO: Make this a config option
    'tsconfig.json'
  )
  const configFile = ts.readConfigFile(configFileName!, ts.sys.readFile)
  const compilerOptions = ts.parseJsonConfigFileContent(
    configFile.config,
    ts.sys,
    './'
  )
  return compilerOptions
}
