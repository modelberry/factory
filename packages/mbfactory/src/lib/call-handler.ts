import chalk from 'chalk'
import { TypeData, PluginData } from './get-all-plugin-data/get-all-plugin-data'
import { ModelberryVariable } from './parse-modelberry-variable'

export type Command = 'push' | 'pull'
export type ArgvType = 'content' | 'models'

export interface CallHandler {
  command: Command
  type: ArgvType
  path?: string
  pluginData?: PluginData
  pluginName: string
}

export type PushHandlerArgs = {
  command: Command
  type: ArgvType
  pluginData?: {
    types: TypeData
    dataVar: ModelberryVariable
  }
}

export type PushHandler = (args: PushHandlerArgs) => Promise<void>

export interface Module {
  handler?: PushHandler
}

export const callHandler = async ({
  command,
  type,
  pluginData,
  pluginName,
}: CallHandler) => {
  const log = console.log
  log(chalk.bold.underline(`\nRunning plugin`))
  console.log(chalk(`- plugin: ${pluginName}`))
  let module: Module = {}
  try {
    module = <Module>await import(pluginName)
  } catch (e) {
    console.log(chalk.red(`- could not find plugin (${e.code})`))
    return
  }

  if (module.handler) {
    await module.handler({ command, type, pluginData })
  } else {
    log(chalk.red(`- could not find handler method on plugin`))
    return
  }
}
