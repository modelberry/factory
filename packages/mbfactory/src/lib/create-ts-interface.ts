import ts from 'typescript'
import { InlineTags } from './create-ts-doc-comment-from-tags'
import { createTsProperty } from './create-ts-property'

export type InterfaceFields = {
  [fieldId: string]: {
    blockTag: string
    inlineTags: InlineTags
    isRequired: boolean
    name: string
  }
}

export type CreateTsInterface = {
  //   blockTag: string
  //   fields: InterfaceFields
  //   inlineTags: InlineTags
  isExported: boolean
  name: string
}

export const createTsInterface = ({
  //   blockTag,
  //   fields,
  //   inlineTags,
  isExported,
  name,
}: CreateTsInterface) => {
  const propertyTree = {
    abstract: { node: { syntaxKind: ts.SyntaxKind.StringKeyword } },
  }
  const members = createTsProperty({ propertyTree })

  const interfaceDeclaration = ts.factory.createInterfaceDeclaration(
    undefined, // Decorators
    isExported
      ? [ts.factory.createToken(ts.SyntaxKind.ExportKeyword)]
      : undefined,
    name,
    undefined, // Type parameters
    undefined, // heritage classes
    members
  )

  ts.addSyntheticLeadingComment(
    interfaceDeclaration,
    ts.SyntaxKind.MultiLineCommentTrivia,
    `* My comment`,
    true
  )
  return interfaceDeclaration
}
