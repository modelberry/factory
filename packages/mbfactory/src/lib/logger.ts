import { inspect } from 'util'
import chalk from 'chalk'
import { Options } from '../call-handler/call-handler'

export interface LoggerInit {
  options: Options
}

export type Logger = {
  init: (args: LoggerInit) => void
  h1: (message: string) => void
  h2: (message: string) => void
  h3: (message: string) => void
  p: (message: string) => void
  info: (message: string) => void
  warning: (message: string) => void
  error: (message: string) => void
  log: (message: string) => void
  object: (heading: string, object: any) => void
  options: Options
  mute: boolean
}

export type LoggerFactory = () => Logger

export const LoggerFactory: LoggerFactory = () => ({
  init({ options }: LoggerInit) {
    Object.assign(this.options, options)
  },
  h1(message: string) {
    this.log(chalk.bold.underline(message))
  },
  h2(message: string) {
    this.log(chalk.bold(message))
  },
  h3(message: string) {
    this.log(chalk.underline(message))
  },
  p(message: string) {
    this.log(chalk(message))
  },
  info(message: string) {
    this.log(chalk.blue(message))
  },
  warning(message: string) {
    this.log(chalk.red.bold(message))
  },
  error(message: string) {
    this.log(chalk.red(message))
  },
  log(message: string) {
    if (this.mute) return
    console.log(message)
  },
  object(heading: string, object: any) {
    if (this.mute) return
    console.log(heading, inspect(object, false, 10, true))
  },
  options: {},
  mute: false,
})

export const logger = LoggerFactory()
