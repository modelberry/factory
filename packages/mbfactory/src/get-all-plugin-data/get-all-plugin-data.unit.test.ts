import ts from 'typescript'
import { getCompilerOptions } from '../get-compiler-options/get-compiler-options'
import { logger } from '../lib/logger'
import { getAllPluginData } from './get-all-plugin-data'

// jest.setTimeout(20000)

const logSpy = {
  h1: jest.spyOn(logger, 'h1').mockImplementation(),
  h2: jest.spyOn(logger, 'h2').mockImplementation(),
  h3: jest.spyOn(logger, 'h3').mockImplementation(),
  p: jest.spyOn(logger, 'p').mockImplementation(),
  info: jest.spyOn(logger, 'info').mockImplementation(),
  error: jest.spyOn(logger, 'error').mockImplementation(),
}

describe('getAllPluginData should', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  const compilerOptions = getCompilerOptions()

  test('get data for topic.ts', async () => {
    const file = './src/get-all-plugin-data/__fixtures__/topic.ts'
    const program = ts.createProgram([file], compilerOptions.options)
    const allPluginData = getAllPluginData({ program })
    // Remove variables that contain hard system paths
    allPluginData['@modelberry/plugin-contentful'].dataVar.fileName=''
    allPluginData['@modelberry/plugin-contentful'].types.ContentfulAction.variables[0].fileName = ''
    allPluginData['@modelberry/plugin-contentful'].types.ContentfulTopic.variables[0].fileName = ''
    expect(allPluginData).toMatchSnapshot()
  })

  test('get data for all-tags.ts', async () => {
    const file = './src/get-all-plugin-data/__fixtures__/all-tags.ts'
    const program = ts.createProgram([file], compilerOptions.options)
    const allPluginData = getAllPluginData({ program })
    expect(allPluginData).toMatchSnapshot()
  })

  test('report processing file for no-tags.ts', async () => {
    const file = './src/get-all-plugin-data/__fixtures__/no-tags.ts'
    const program = ts.createProgram([file], compilerOptions.options)
    getAllPluginData({ program })
    expect(logSpy.h3).toHaveBeenCalledWith(
      'src/get-all-plugin-data/__fixtures__/no-tags.ts'
    )
  })

  test('report warnings for no-tags.ts (parse-wr-interface)', async () => {
    const file = './src/get-all-plugin-data/__fixtures__/no-tags.ts'
    const program = ts.createProgram([file], compilerOptions.options)
    getAllPluginData({ program })
    expect(logSpy.error).toHaveBeenCalledWith('- no TSDoc tags (Topic)')
  })

  test('report warnings for no-plugin-tag (parse-wr-interface)', async () => {
    const file = './src/get-all-plugin-data/__fixtures__/no-plugin-tag'
    const program = ts.createProgram([file], compilerOptions.options)
    getAllPluginData({ program })
    expect(logSpy.error).toHaveBeenCalledWith('- no @plugin inline tag (Topic)')
  })

  test('report warnings for no-modelberry-field-tag (parse-wr-interface)', async () => {
    const file =
      './src/get-all-plugin-data/__fixtures__/no-modelberry-field-tag'
    const program = ts.createProgram([file], compilerOptions.options)
    getAllPluginData({ program })
    expect(logSpy.error).toHaveBeenCalledWith(
      '- no @modelberry block tag for field (Topic/heading)'
    )
  })

  test('report warnings for no-modelberry-interface-tag (parse-wr-interface)', async () => {
    const file =
      './src/get-all-plugin-data/__fixtures__/no-modelberry-interface-tag'
    const program = ts.createProgram([file], compilerOptions.options)
    getAllPluginData({ program })
    expect(logSpy.error).toHaveBeenCalledWith(
      '- no @modelberry block tag for interface (Topic)'
    )
  })
})
