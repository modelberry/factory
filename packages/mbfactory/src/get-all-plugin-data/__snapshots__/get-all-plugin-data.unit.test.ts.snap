// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllPluginData should get data for all-tags.ts 1`] = `
{
  "@modelberry/plugin-contentful": {
    "dataVar": {},
    "types": {
      "Topic": {
        "interface": {
          "fields": {
            "heading": {
              "tags": {
                "@helpText": "The topic heading",
                "@itemsLinkType": "Entry",
                "@itemsType": "Link",
                "@linkType": "Symbol",
                "@localized": "",
                "@name": "Heading",
                "@required": "",
                "@type": "Array",
                "@validations": "unique",
                "@widgetId": "singleLine",
              },
              "type": "string",
            },
          },
          "interfaceTags": {
            "@plugin": "@modelberry/plugin-contentful",
            "@type": "topic",
          },
          "typeName": "Topic",
        },
        "variables": [],
      },
    },
  },
}
`;

exports[`getAllPluginData should get data for topic.ts 1`] = `
{
  "@modelberry/plugin-contentful": {
    "dataVar": {
      "fileName": "",
      "isArray": false,
      "isExported": true,
      "isObject": true,
      "isTyped": true,
      "name": "modelberryPluginData",
      "type": "ModelberryPluginData",
      "value": "{
  '@modelberry/plugin-contentful': {
    validations: {
      dropdown: { in: ['item A', 'item B', 'item C'] },
      media: { linkMimetypeGroup: ['image', 'video'] },
      action: { linkContentType: ['testAction'] },
      shortString: { size: { max: 155, min: 0 } },
      camelCase: {
        regexp: { pattern: '^[a-z]+([A-Z][a-z0-9]+)*$', flags: '' },
      },
      path: {
        regexp: {
          pattern: '^\\\\/[a-z0-9:\\\\._/~%\\\\-\\\\+&\\\\#\\\\?!=\\\\(\\\\)@]*$',
          flags: '',
        },
      },
      url: {
        regexp: {
          pattern:
            '^(ftp|http|https|mailto):(\\\\/\\\\/)*(\\\\w+:{0,1}\\\\w*@)?(\\\\S+)(:[0-9]+)?(\\\\/|\\\\/([\\\\w#!:.?+=&%@!\\\\-/]))?$',
          flags: '',
        },
      },
    },
  },
}",
    },
    "types": {
      "ContentfulAction": {
        "interface": {
          "fields": {
            "anchor": {
              "tags": {
                "@helpText": "Voorbeeld: mijn-anchor",
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
            "description": {
              "tags": {
                "@helpText": "Omschrijving van de actie, voor zoekmachines",
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
            "eventId": {
              "tags": {
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
            "heading": {
              "tags": {
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
            "query": {
              "tags": {
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
            "url": {
              "tags": {
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
          },
          "interfaceTags": {
            "@displayField": "heading",
            "@plugin": "@modelberry/plugin-contentful",
            "@type": "testAction",
          },
          "typeName": "ContentfulAction",
        },
        "variables": [
          {
            "fileName": "",
            "isArray": true,
            "isExported": true,
            "isObject": false,
            "isTyped": true,
            "name": "myActions",
            "type": "ContentfulAction",
            "value": "[
  {
    sys: {
      id: 'myContentActionId',
    },
    heading: 'Hello',
    anchor: 'anchor',
  },
]",
          },
        ],
      },
      "ContentfulTopic": {
        "interface": {
          "fields": {
            "abstract": {
              "tags": {
                "@type": "Symbol",
                "@validations": "shortString",
              },
              "type": "string | undefined",
            },
            "actionsCollection": {
              "tags": {
                "@description": "Action field, this is the description",
                "@itemsLinkType": "Entry",
                "@itemsType": "Link",
                "@type": "Array",
                "@validations": "action",
              },
              "type": "{ items: ContentfulAction[]; } | undefined",
            },
            "heading": {
              "tags": {
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
            "icon": {
              "tags": {
                "@type": "Symbol",
              },
              "type": "string | undefined",
            },
            "mediaCollection": {
              "tags": {
                "@itemsLinkType": "Asset",
                "@itemsType": "Link",
                "@type": "Array",
              },
              "type": "{ items: ContentfulAsset[]; } | undefined",
            },
            "poster": {
              "tags": {
                "@linkType": "Asset",
                "@type": "Link",
              },
              "type": "ContentfulAsset | undefined",
            },
          },
          "interfaceTags": {
            "@description": "Topic model, a heading, an abstract and a call to action",
            "@displayField": "heading",
            "@plugin": "@modelberry/plugin-contentful",
            "@type": "testTopic",
          },
          "typeName": "ContentfulTopic",
        },
        "variables": [
          {
            "fileName": "",
            "isArray": true,
            "isExported": true,
            "isObject": false,
            "isTyped": true,
            "name": "myTopics",
            "type": "ContentfulTopic",
            "value": "[
  {
    abstract: 'This is the abstract',
    actionsCollection: {
      items: [
        {
          sys: {
            id: 'myContentActionId',
          },
        },
      ],
    },
    heading: 'Heres the heading',
  },
]",
          },
        ],
      },
    },
  },
}
`;
